packages/amazonq/src/app/inline/recommendationService.ts

```diff
@@ -119,7 +122,51 @@ export class RecommendationService {
             })
             const t0 = performance.now()
 
-            const result = await this.getRecommendationsWithTimeout(languageClient, request, token)
+            // Best effort estimate of deletion
+            const isTriggerByDeletion = documentEventListener.isLastEventDeletion(document.uri.fsPath)
+
+            const ps: Promise<InlineCompletionListWithReferences>[] = []
+            /**
+             * IsTriggerByDeletion is to prevent user deletion invoking Completions.
+             * PartialResultToken is not a hack for now since only Edits suggestion use partialResultToken across different calls of [getAllRecommendations],
+             * Completions use PartialResultToken with single 1 call of [getAllRecommendations].
+             * Edits leverage partialResultToken to achieve EditStreak such that clients can pull all continuous suggestions generated by the model within 1 EOS block.
+             */
+            if (!isTriggerByDeletion && !request.partialResultToken) {
+                const completionPromise: Promise<InlineCompletionListWithReferences> = languageClient.sendRequest(
+                    inlineCompletionWithReferencesRequestType.method,
+                    request,
+                    token
+                )
+                ps.push(completionPromise)
+            }
+
+            /**
+             * Though Edit request is sent on keystrokes everytime, the language server will execute the request in a debounced manner so that it won't be immediately executed.
+             */
+            const editPromise: Promise<InlineCompletionListWithReferences> = languageClient.sendRequest(
+                editCompletionRequestType.method,
+                request,
+                token
+            )
+            ps.push(editPromise)
+
+            /**
+             * First come first serve, ideally we should simply return the first response returned. However there are some caviar here because either
+             * (1) promise might be returned early without going through service
+             * (2) some users are not enabled with edits suggestion, therefore service will return empty result without passing through the model
+             * With the scenarios listed above or others, it's possible that 1 promise will ALWAYS win the race and users will NOT get any suggestion back.
+             * This is the hack to return first "NON-EMPTY" response
+             */
+            let result = await Promise.race(ps)
+            if (ps.length > 1 && result.items.length === 0) {
+                for (const p of ps) {
+                    const r = await p
+                    if (r.items.length > 0) {
+                        result = r
+                    }
+                }
+            }
```

Comment:
Ugly, but this is the best effort we can do atm

Reply:
We should add another field `suggestionType` within `InlineCompletionListWithReferences` so that we know which suggestion type it is instead of this hack. Can follow up.